File: sdk/go/channel_domains.go
package tempemail

import "strings"

// channelDomains 维护渠道到已知支持域名的静态映射
var channelDomains = map[Channel][]string{
	// Gmail 相关
	ChannelEmailnator:  {"gmail.com", "googlemail.com"},
	ChannelTempgmailer: {"gmail.com"},

	// 多域名渠道
	ChannelCatchmail:             {"catchmail.io", "mailistry.com", "zeppost.com"},
	ChannelCatchmailMailistry:    {"mailistry.com"},
	ChannelCatchmailZeppost:      {"zeppost.com"},
	ChannelMailforspam:           {"mailforspam.com", "tempmail.io", "disposable.email"},
	ChannelMailforspamTempmailIo: {"tempmail.io"},
	ChannelMailforspamDisposable: {"disposable.email"},
	ChannelTempmail365:           {"fengyou.cc", "shop345.com", "nutemail.com", "qvrf.cn"},
	ChannelTempinbox:             {"tempinbox.xyz", "thepiratebay.cloud", "cryptoblad.nl"},
	Channel10minuteOne:           {"xghff.com", "oqqaj.com", "psovv.com", "dbwot.com", "ygwpr.com", "imxwe.com"},
	ChannelXghffCom:              {"xghff.com"},
	ChannelOqqajCom:              {"oqqaj.com"},
	ChannelPsovvCom:              {"psovv.com"},
	ChannelDbwotCom:              {"dbwot.com"},
	ChannelYgwprCom:              {"ygwpr.com"},
	ChannelImxweCom:              {"imxwe.com"},
	ChannelApihz:                 {"apimail.email", "apimail.vip"},
	Channel24mailChacuo:          {"chacuo.net", "027168.com"},

	// mailinator 系列
	ChannelMailinator:              {"mailinator.com"},
	ChannelSogetthisCom:            {"sogetthis.com"},
	ChannelBobmailInfo:             {"bobmail.info"},
	ChannelSuremailInfo:            {"suremail.info"},
	ChannelBinkmailCom:             {"binkmail.com"},
	ChannelVeryrealemailCom:        {"veryrealemail.com"},
	ChannelChammyInfo:              {"chammy.info"},
	ChannelThisisnotmyrealemailCom: {"thisisnotmyrealemail.com"},
	ChannelNotmailinatorCom:        {"notmailinator.com"},
	ChannelSpamherepleaseCom:       {"spamhereplease.com"},
	ChannelSendspamhereCom:         {"sendspamhere.com"},
	ChannelSendfreeOrg:             {"sendfree.org"},

	// 其他单域名
	ChannelMailnesia:    {"mailnesia.com"},
	ChannelHarakirimail: {"harakirimail.com"},
	ChannelByom:         {"byom.de"},
	ChannelEyepaste:     {"eyepaste.com"},
	ChannelMailcatch:    {"mailcatch.com"},
	ChannelMaildropCc:   {"maildrop.cc"},
	ChannelSmailPw:      {"smail.pw"},
	ChannelInboxkitten:  {"inboxkitten.com"},
	ChannelMffac:        {"mffac.com"},
	ChannelNimail:       {"nimail.cn"},
	ChannelMailgolem:    {"mailgolem.com"},
	ChannelMohmal:       {"emailinbo.live"},

	// tempmail-plus 系列
	ChannelTempmailPlus: {"mailto.plus"},
	ChannelFexpostCom:   {"fexpost.com"},
	ChannelFexboxOrg:    {"fexbox.org"},
	ChannelMailboxInUa:  {"mailbox.in.ua"},
	ChannelRoverInfo:    {"rover.info"},
	ChannelChitthiIn:    {"chitthi.in"},
	ChannelFextempCom:   {"fextemp.com"},
	ChannelAnyPink:      {"any.pink"},
	ChannelMerepostCom:  {"merepost.com"},

	// getnada 系列
	ChannelGetnada:      {"getnada.com", "getnada.email", "getnada.net"},
	ChannelGetnadaCom:   {"getnada.com"},
	ChannelGetnadaEmail: {"getnada.email"},
	ChannelGetnadaNet:   {"getnada.net"},

	// disposablemail
	ChannelDisposablemailApp: {"disposablemail.dev", "mailmehere.cc"},

	// minuteinbox
	ChannelMinuteinbox: {"minafter.com"},

	// xkx.me
	ChannelXkxMe: {"xkx.me"},
}

// dynamicDomainChannels 是域名通过 API 动态获取的渠道
// 在域名筛选时默认保留这些渠道（无法预判其支持的域名）
var dynamicDomainChannels = map[Channel]bool{
	ChannelMailTm:        true,
	ChannelDuckmail:      true,
	ChannelMailTd:        true,
	ChannelChatgptOrgUk:  true,
	ChannelTemporam:      true,
	ChannelGetnada:       true,
	ChannelNeighbours:    true,
	ChannelNeighboursSh:  true,
	ChannelInboxes:       true,
	ChannelMailmomy:      true,
	ChannelFreecustom:    true,
	ChannelMailSunls:     true,
	ChannelAnonymmail:    true,
	ChannelFakeLegal:     true,
	ChannelMaildrop:      true,
	ChannelMailCx:        true,
	ChannelMoakt:         true,
	ChannelTempmailLol:   true,
	ChannelSmailpro:      true,
	ChannelTempmail:      true,
	ChannelOneSecMail:    true,
	ChannelGuerrillaMail: true,
	ChannelDropmail:      true,
	ChannelFmail:         true,
	ChannelOckito:        true,
	ChannelTempMailOrg:   true,
}

// filterChannelsByDomain 根据目标域名过滤渠道列表
// targetDomains 为用户指定的域名列表
// 返回过滤后的渠道列表；如果 targetDomains 为空则返回原列表
func filterChannelsByDomain(channels []Channel, targetDomains []string) []Channel {
	if len(targetDomains) == 0 {
		return channels
	}

	var filtered []Channel
	for _, ch := range channels {
		// 动态域名渠道默认保留
		if dynamicDomainChannels[ch] {
			filtered = append(filtered, ch)
			continue
		}
		// 检查该渠道是否支持目标域名
		domains, ok := channelDomains[ch]
		if !ok {
			continue
		}
		if matchesDomain(domains, targetDomains) {
			filtered = append(filtered, ch)
		}
	}

	// 如果过滤后为空，回退到原列表
	if len(filtered) == 0 {
		return channels
	}
	return filtered
}

// matchesDomain 检查渠道支持的域名列表是否包含任一目标域名
func matchesDomain(chDomains []string, targetDomains []string) bool {
	for _, cd := range chDomains {
		for _, td := range targetDomains {
			// 精确匹配或子域名匹配
			if cd == td || strings.HasSuffix(cd, "."+td) {
				return true
			}
		}
	}
	return false
}

File: sdk/rust/src/channel_domains.rs
/*!
 * 渠道域名映射模块
 * 用于根据邮箱后缀/域名筛选可用渠道
 */

use std::collections::{HashMap, HashSet};
use std::sync::LazyLock;

use crate::types::Channel;

/// 渠道到已知固定域名的映射表
static CHANNEL_DOMAINS: LazyLock<HashMap<Channel, Vec<&'static str>>> = LazyLock::new(|| {
    let mut m: HashMap<Channel, Vec<&'static str>> = HashMap::new();
    m.insert(Channel::Emailnator, vec!["gmail.com", "googlemail.com"]);
    m.insert(Channel::Tempgmailer, vec!["gmail.com"]);
    m.insert(Channel::Catchmail, vec!["catchmail.io", "mailistry.com", "zeppost.com"]);
    m.insert(Channel::CatchmailMailistry, vec!["mailistry.com"]);
    m.insert(Channel::CatchmailZeppost, vec!["zeppost.com"]);
    m.insert(Channel::Mailforspam, vec!["mailforspam.com", "tempmail.io", "disposable.email"]);
    m.insert(Channel::MailforspamTempmailIo, vec!["tempmail.io"]);
    m.insert(Channel::MailforspamDisposable, vec!["disposable.email"]);
    m.insert(Channel::Tempmail365, vec!["fengyou.cc", "shop345.com", "nutemail.com", "qvrf.cn"]);
    m.insert(Channel::Tempinbox, vec!["tempinbox.xyz", "thepiratebay.cloud", "cryptoblad.nl"]);
    m.insert(Channel::TenminuteOne, vec!["xghff.com", "oqqaj.com", "psovv.com", "dbwot.com", "ygwpr.com", "imxwe.com"]);
    m.insert(Channel::XghffCom, vec!["xghff.com"]);
    m.insert(Channel::OqqajCom, vec!["oqqaj.com"]);
    m.insert(Channel::PsovvCom, vec!["psovv.com"]);
    m.insert(Channel::DbwotCom, vec!["dbwot.com"]);
    m.insert(Channel::YgwprCom, vec!["ygwpr.com"]);
    m.insert(Channel::ImxweCom, vec!["imxwe.com"]);
    m.insert(Channel::Apihz, vec!["apimail.email", "apimail.vip"]);
    m.insert(Channel::TwentyfourmailChacuo, vec!["chacuo.net", "027168.com"]);
    m.insert(Channel::Mailinator, vec!["mailinator.com"]);
    m.insert(Channel::SogetthisCom, vec!["sogetthis.com"]);
    m.insert(Channel::BobmailInfo, vec!["bobmail.info"]);
    m.insert(Channel::SuremailInfo, vec!["suremail.info"]);
    m.insert(Channel::BinkmailCom, vec!["binkmail.com"]);
    m.insert(Channel::VeryrealemailCom, vec!["veryrealemail.com"]);
    m.insert(Channel::ChammyInfo, vec!["chammy.info"]);
    m.insert(Channel::ThisisnotmyrealemailCom, vec!["thisisnotmyrealemail.com"]);
    m.insert(Channel::NotmailinatorCom, vec!["notmailinator.com"]);
    m.insert(Channel::SpamherepleaseCom, vec!["spamhereplease.com"]);
    m.insert(Channel::SendspamhereCom, vec!["sendspamhere.com"]);
    m.insert(Channel::SendfreeOrg, vec!["sendfree.org"]);
    m.insert(Channel::Mailnesia, vec!["mailnesia.com"]);
    m.insert(Channel::Harakirimail, vec!["harakirimail.com"]);
    m.insert(Channel::Byom, vec!["byom.de"]);
    m.insert(Channel::Eyepaste, vec!["eyepaste.com"]);
    m.insert(Channel::Mailcatch, vec!["mailcatch.com"]);
    m.insert(Channel::MaildropCc, vec!["maildrop.cc"]);
    m.insert(Channel::SmailPw, vec!["smail.pw"]);
    m.insert(Channel::Inboxkitten, vec!["inboxkitten.com"]);
    m.insert(Channel::Mffac, vec!["mffac.com"]);
    m.insert(Channel::Nimail, vec!["nimail.cn"]);
    m.insert(Channel::Mailgolem, vec!["mailgolem.com"]);
    m.insert(Channel::TempmailPlus, vec!["mailto.plus"]);
    m.insert(Channel::FexpostCom, vec!["fexpost.com"]);
    m.insert(Channel::FexboxOrg, vec!["fexbox.org"]);
    m.insert(Channel::MailboxInUa, vec!["mailbox.in.ua"]);
    m.insert(Channel::RoverInfo, vec!["rover.info"]);
    m.insert(Channel::ChitthiIn, vec!["chitthi.in"]);
    m.insert(Channel::FextempCom, vec!["fextemp.com"]);
    m.insert(Channel::AnyPink, vec!["any.pink"]);
    m.insert(Channel::MerepostCom, vec!["merepost.com"]);
    m.insert(Channel::Getnada, vec!["getnada.com"]);
    m.insert(Channel::GetnadaCom, vec!["getnada.com"]);
    m.insert(Channel::GetnadaEmail, vec!["getnada.email"]);
    m.insert(Channel::GetnadaNet, vec!["getnada.net"]);
    m.insert(Channel::DisposablemailApp, vec!["disposablemail.dev", "mailmehere.cc"]);
    m.insert(Channel::Minuteinbox, vec!["minafter.com"]);
    m.insert(Channel::Mohmal, vec!["emailinbo.live"]);
    m
});

/// 动态域名渠道集合（这些渠道的域名列表是动态获取的，筛选时默认保留）
static DYNAMIC_DOMAIN_CHANNELS: LazyLock<HashSet<Channel>> = LazyLock::new(|| {
    let mut s = HashSet::new();
    s.insert(Channel::MailTm);
    s.insert(Channel::Duckmail);
    s.insert(Channel::MailTd);
    s.insert(Channel::ChatgptOrgUk);
    s.insert(Channel::Temporam);
    s.insert(Channel::Getnada);
    s.insert(Channel::Neighbours);
    s.insert(Channel::NeighboursSh);
    s.insert(Channel::Inboxes);
    s.insert(Channel::Mailmomy);
    s.insert(Channel::Freecustom);
    s.insert(Channel::MailSunls);
    s.insert(Channel::Anonymmail);
    s.insert(Channel::FakeLegal);
    s.insert(Channel::Maildrop);
    s.insert(Channel::MailCx);
    s.insert(Channel::Moakt);
    s.insert(Channel::TempmailLol);
    s.insert(Channel::Smailpro);
    s.insert(Channel::Tempmail);
    s.insert(Channel::OneSecMail);
    s.insert(Channel::GuerrillaMail);
    s.insert(Channel::Dropmail);
    s.insert(Channel::Fmail);
    s.insert(Channel::Ockito);
    s.insert(Channel::TempMailOrg);
    s
});

/// 根据目标域名筛选渠道列表
///
/// 如果 target_domains 为空，返回原渠道列表。
/// 动态域名渠道默认保留（因为无法提前确定其支持的域名）。
/// 如果筛选结果为空，返回原渠道列表（避免完全无法使用）。
pub fn filter_channels_by_domain(channels: &[Channel], target_domains: &[String]) -> Vec<Channel> {
    if target_domains.is_empty() {
        return channels.to_vec();
    }

    let filtered: Vec<Channel> = channels
        .iter()
        .filter(|ch| {
            // 动态域名渠道默认保留
            if DYNAMIC_DOMAIN_CHANNELS.contains(ch) {
                return true;
            }
            // 检查渠道已知域名是否与目标匹配
            if let Some(domains) = CHANNEL_DOMAINS.get(ch) {
                return domains.iter().any(|cd| {
                    target_domains
                        .iter()
                        .any(|td| *cd == td.as_str() || cd.ends_with(&format!(".{}", td)))
                });
            }
            false
        })
        .cloned()
        .collect();

    // 筛选结果为空时回退到原列表
    if filtered.is_empty() {
        channels.to_vec()
    } else {
        filtered
    }
}
